src/crimson/os/seastore/onode_manager/staged-fltree/stages/key_layout.h

```diff
@@ -48,7 +48,7 @@ static laddr_t get_lba_hint(shard_t shard, pool_t pool, crush_hash_t crush) {
   // conflicts.
   return ((uint64_t)(shard & 0XFF)<<56 |
           (uint64_t)(pool  & 0xFF)<<48 |
-          (uint64_t)(crush       )<<16);
+          (uint64_t)(crush & (((uint64_t)0x1 << (64 - ONODE_LADDR_SPACE_ORDER)) - 1))<<32);
```

Comment:
```
crush & (((uint64_t)0x1 << (64 - ONODE_LADDR_SPACE_ORDER)) - 1)
```
This seems to be masking a 32 bit value with ((1<<32) - 1), 0xFFFFFFFF?

I think we can make this even simpler, just modify get_lba_hint here to return a uniform random value:

```
+constexpr uint32_t ONODE_HINT_RANGE = 32<<20; // Randomize hints over 32MB
 static laddr_t get_lba_hint(shard_t shard, pool_t pool, crush_hash_t crush) {
   // FIXME: It is possible that PGs from different pools share the same prefix
   // if the mask 0xFF is not long enough, result in unexpected transaction
   // conflicts.
-  return ((uint64_t)(shard & 0XFF)<<56 |
-          (uint64_t)(pool  & 0xFF)<<48 |
-          (uint64_t)(crush       )<<16);
+  auto hint = ((uint64_t)(shard & 0XFF)<<56 |
+              (uint64_t)(pool  & 0xFF)<<48 |
+              (uint64_t)(crush       )<<16);
+  // Randomize hints to avoid addresses for a particular object clustering
+  // at the hint.
+  return hint + (std::rand() % ONODE_HINT_RANGE);
 }
```

Reply:
> This seems to be masking a 32 bit value with ((1<<32) - 1), 0xFFFFFFFF?

Oh, sorry, this is a mistake, should be 1<<32

> I think we can make this even simpler, just modify get_lba_hint here to return a uniform random value:
> 
> ```
> +constexpr uint32_t ONODE_HINT_RANGE = 32<<20; // Randomize hints over 32MB
>  static laddr_t get_lba_hint(shard_t shard, pool_t pool, crush_hash_t crush) {
>    // FIXME: It is possible that PGs from different pools share the same prefix
>    // if the mask 0xFF is not long enough, result in unexpected transaction
>    // conflicts.
> -  return ((uint64_t)(shard & 0XFF)<<56 |
> -          (uint64_t)(pool  & 0xFF)<<48 |
> -          (uint64_t)(crush       )<<16);
> +  auto hint = ((uint64_t)(shard & 0XFF)<<56 |
> +              (uint64_t)(pool  & 0xFF)<<48 |
> +              (uint64_t)(crush       )<<16);
> +  // Randomize hints to avoid addresses for a particular object clustering
> +  // at the hint.
> +  return hint + (std::rand() % ONODE_HINT_RANGE);
>  }
> ```

I think `crush << 16` leaves only 64KB between onodes, so I used `(crush & (1 << 32)) << 32` to enlarge that interval to 4GB, so onodes are unlikely to collide with each other.

Reply:
> I think crush << 16 leaves only 64KB between onodes

`crush << 16` randomizes onode locations at the 48-bit range addresses because `crush` is a hashed value.

Reply:
> I think we can make this even simpler, just modify get_lba_hint here to return a uniform random value:
> 
> ```
> +constexpr uint32_t ONODE_HINT_RANGE = 32<<20; // Randomize hints over 32MB
>  static laddr_t get_lba_hint(shard_t shard, pool_t pool, crush_hash_t crush) {
>    // FIXME: It is possible that PGs from different pools share the same prefix
>    // if the mask 0xFF is not long enough, result in unexpected transaction
>    // conflicts.
> -  return ((uint64_t)(shard & 0XFF)<<56 |
> -          (uint64_t)(pool  & 0xFF)<<48 |
> -          (uint64_t)(crush       )<<16);
> +  auto hint = ((uint64_t)(shard & 0XFF)<<56 |
> +              (uint64_t)(pool  & 0xFF)<<48 |
> +              (uint64_t)(crush       )<<16);
> +  // Randomize hints to avoid addresses for a particular object clustering
> +  // at the hint.
> +  return hint + (std::rand() % ONODE_HINT_RANGE);
>  }
> ```

I don't quite follow, do you mean that we randomize onode data laddr start point and omap laddr hints within the 32MB space?

I think this would mean that half of omap extents laddr hints will collide with onode data laddrs, because onode data is 16MB long, occupying half of the 32MB space. Am I right?

Reply:
PG log objects don't have data payloads, and so do not have a 16MB reservation.

Reply:
Probably `constexpr uint32_t ONODE_HINT_RANGE = 1<<30 ; //Randomize hints over 1GB` for now?

This also means collision over 2^18 onodes per pool, should be enough?

Reply:
This doesn't actually work, the omap implementation caches a single hint value.  Needs a bit more of a refactor.

Reply:
> PG log objects don't have data payloads, and so do not have a 16MB reservation.

ok, will do:-)

Reply:
Objects don't typically have 32MB of data.  Randomizing over 1GB would tend to cause blocks from adjacent objects to overlap to a larger degree.  We could make it a config and test different values.

Reply:
> Huh, I guess omap_root_t::hint doesn't actually get persisted as part of omap_root_le_t, it gets set fresh each time we read it. This should work fine.

It seems that omap_root_t::hint will stay the same until the next `BtreeOMapManager::initialize_omap` is called again, am I right? If so, omap extents' hint will still have a high possibility of collision, am I right?

Reply:
> Objects don't typically have 32MB of data. Randomizing over 1GB would tend to cause blocks from adjacent objects to overlap to a larger degree. We could make it a config and test different values.

will do:-)

Reply:
https://github.com/athanatos/ceph/commit/5ec5bd54d5f3c692cf09ce7e5939477ab568b390 is I think a bit cleaner.  It updates the hint mechanism in Onode to provide different data and metadata hints and updates the callers appropriately.

Reply:
> It seems that omap_root_t::hint will stay the same until the next `BtreeOMapManager::initialize_omap` is called again, am I right? If so, omap extents' hint will still have a high possibility of collision, am I right? We still have to further randomize omap extent laddr hints even if omap_root_t::hint doesn't change, right?

It's confusing, but no.  omap_root_le_t doesn't actually store the hint.  omap_root_le_t::get(laddr_t hint) populates the returned omap_root_t with the passed hint each time it's invoked.  Really, hint shouldn't be part of omap_root_t.

Reply:
> [athanatos@5ec5bd5](https://github.com/athanatos/ceph/commit/5ec5bd54d5f3c692cf09ce7e5939477ab568b390) is I think a bit cleaner. It updates the hint mechanism in Onode to provide different data and metadata hints and updates the callers appropriately. It does ensure that metadata and data hints won't overlap as well in case we end up with objects that use both.

Ah, finally, I see:-) Will cherry-pick that commit, and repush the pr:-)